disable multi_linear_share_same_input for dynamic shape case

zhuhaozhe opened the pull request:
`reshape_linear_reshape_pattern` will fail for dynamic shapes and break the fusion.
We will disable this optimization for dynamic shapes, since the shape might be changed during runtime so we cannot compare the size hint.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107123



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

zhuhaozhe committed:
disable multi_linear_share_same_input for dynamic shape case

[ghstack-poisoned]

zhuhaozhe committed:
Update on "disable multi_linear_share_same_input for dynamic shape case"

`reshape_linear_reshape_pattern` will fail for dynamic shapes and break the fusion.
We will disable this optimization for dynamic shapes, since the shape might be changed during runtime so we cannot compare the size hint.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

zhuhaozhe committed:
Update on "disable multi_linear_share_same_input for dynamic shape case"

`reshape_linear_reshape_pattern` will fail for dynamic shapes and break the fusion.
We will disable this optimization for dynamic shapes, since the shape might be changed during runtime so we cannot compare the size hint.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -655,6 +655,15 @@ def _recover_linear():
+                isinstance(x, int) for x in ([reshape_1[0]] + reshape_2[:-1])
+            )
+            if dynamic_shapes:
+                # For dynamic shapes, we are unsafe to compair the shapes for reshape_1 and reshape_2
+                # since they can be changed during runtime
```
Can we add guard to also support dynamic shape?

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -655,6 +655,15 @@ def _recover_linear():
+
             graph = match.graph
             node = match.output_node()
             if reshape_1[0] == reduce(lambda x, y: x * y, reshape_2[:-1]):
```
How does `mkldnn._linear_pointwise.default` produce the output of the same shape as the one from `reshape_2`?

zhuhaozhe replied:
Also checks
`linear's input's shape[:-1] == reshape_2[:-1]` and `product(reshape_2[:-1]) == reshape_1[0]`

zhuhaozhe committed:
Update on "disable multi_linear_share_same_input for dynamic shape case"

`reshape_linear_reshape_pattern` will fail for dynamic shapes and break the fusion.
We will disable this optimization for dynamic shapes, since the shape might be changed during runtime so we cannot compare the size hint.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -655,15 +655,36 @@ def _recover_linear():
+                can_remove_reshape = linear_input_node.meta.get("val").shape[
+                    :-1
+                ] == torch.Size(reshape_2[:-1])
+                can_remove_reshape == can_remove_reshape and (
+                    reduce(lambda x, y: x * y, reshape_2[:-1]) == reshape_1[0]
+                )
```
This doesn't seem right.

zhuhaozhe replied:
Thanks for reminder, updated.

zhuhaozhe committed:
Update on "disable multi_linear_share_same_input for dynamic shape case"

`reshape_linear_reshape_pattern` will fail for dynamic shapes and break the fusion.
We will disable this optimization for dynamic shapes, since the shape might be changed during runtime so we cannot compare the size hint.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

zhuhaozhe committed:
Update on "disable multi_linear_share_same_input for dynamic shape case"

`reshape_linear_reshape_pattern` will fail for dynamic shapes and break the fusion.
We will disable this optimization for dynamic shapes, since the shape might be changed during runtime so we cannot compare the size hint.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -655,17 +655,36 @@ def _recover_linear():
+            linear_input_node = reshape_2_node.args[0].args[0].args[0]
+            # check linear's input's shape[:-1] == reshape_2[:-1]
+            # and check product(reshape_2[:-1]) == reshape_1[0]
+            if dynamic_shapes:
+                # TODO: Haozhe investigate how add guard here
+                return
+            else:
+                can_remove_reshape = linear_input_node.meta.get("val").shape[
+                    :-1
+                ] == torch.Size(reshape_2[:-1])
+                can_remove_reshape = can_remove_reshape and (
+                    reduce(lambda x, y: x * y, reshape_2[:-1]) == reshape_1[0]
+                )
```
How about this. It doesn't introduce a guard but can still handle simple symbolically equivalent cases:
```python
            can_remove_reshape = (
                V.graph.sizevars.statically_known_list_equals(
                    linear_input_node.meta.get("val").shape[:-1],
                    reshape_2[:-1]
                )
                and V.graph.sizevars.statically_known_equals(
                    sympy_product(reshape_2[:-1], reshape_1[0]
                )
            )
```

(Assuming V.graph is available at this point?)

zhuhaozhe replied:
Hi, @peterbell10. `V.graph` is not available at this point.
I tried some code from `shape_env.evaulte_expr`, but does not quite understand it yet.
And since it will bring obvious performance regression for https://github.com/pytorch/pytorch/pull/106300.
Can we first disable this for dynamic case?

peterbell10 replied:
Yeah disabling is fine, just thought this would be nicer. If you have a `shape_env` then you could use `shape_env._maybe_evaluate_static` directly?

zhuhaozhe replied:
Thanks for the advice, I will continue to investigate this

zhuhaozhe commented:
@pytorchbot merge
